fastlane_bot/modes/base_triangle.py

```diff
@@ -9,301 +9,83 @@
-            flt_triangle_analysis_set += list(zip([flt] * len(triangles_to_run), triangles_to_run))
-        
-        self.ConfigObj.logger.debug(f"[base_triangle.get_analysis_set_per_flt] Length of flt_triangle_analysis_set: {flt, len(flt_triangle_analysis_set)}")
-        return flt_triangle_analysis_set
-
-    def get_comprehensive_triangles(
-        self, flashloan_tokens: List[str], CCm: Any
-    ) -> Tuple[List[str], List[Any]]:
-        """
-        Get comprehensive combos for triangular arbitrage
+    @abc.abstractmethod
+    def get_combos(self, flashloan_tokens: List[str], CCm: Any, arb_mode: str) -> Tuple[List[str], List[Any]]:
+        ...
 
-        Parameters
-        ----------
-        flashloan_tokens : list
-            List of flashloan tokens
-        CCm : object
-            CCm object
+    def find_arbitrage(self, candidates: List[Any] = None, ops: Tuple = None, best_profit: float = 0, profit_src: float = 0) -> Union[List, Tuple]:
+        self.assert_supported()
 
-        Returns
-        -------
-        combos : list
-            List of combos
+        if candidates is None:
+            candidates = []
 
-        """
-        combos = []
-        for flt in flashloan_tokens:
+        combos = self.get_combos(self.flashloan_tokens, self.CCm, arb_mode=self.arb_mode)
 
-            # Get the Carbon pairs
-            carbon_pairs = sort_pairs(set([x.pair for x in CCm.curves if x.params.exchange in self.ConfigObj.CARBON_V1_FORKS]))
-            
-            # Create a set of unique tokens, excluding 'flt'
-            x_tokens = {token for pair in carbon_pairs for token in pair.split('/') if token != flt}
-            
-            # Get relevant pairs containing the flashloan token
-            flt_x_pairs = sort_pairs([f"{x}/{flt}" for x in x_tokens])
-            
-            # Generate all possible 2-item combinations from the unique tokens that arent the flashloan token
-            x_y_pairs = sort_pairs(["{}/{}".format(x, y) for x, y in itertools.combinations(x_tokens, 2)])
-            
-            # Note the relevant pairs
-            all_relevant_pairs = flt_x_pairs + x_y_pairs
-            self.ConfigObj.logger.debug(f"len(all_relevant_pairs) {len(all_relevant_pairs)}")
+        for src_token, miniverse in combos:
+            try:
+                CC_cc = CPCContainer(miniverse)
+                O = MargPOptimizer(CC_cc)
+                pstart = self.build_pstart(CC_cc, CC_cc.tokens(), src_token)
+                r = O.optimize(src_token, params=dict(pstart=pstart))
+                trade_instructions_dic = r.trade_instructions(O.TIF_DICTS)
+                if trade_instructions_dic is None or len(trade_instructions_dic) < 3:
+                    # Failed to converge
+                    continue
+                trade_instructions_df = r.trade_instructions(O.TIF_DFAGGR)
+                trade_instructions = r.trade_instructions()
```

Comment:
why do you get the trade instructions in all those different formats? they are equivalent and you should choose one

Reply:
Outside the scope of this PR (same code used prior to it)

Reply:
but are they all used below?

Reply:
Annoying indeed.
I was initially aiming for the least amount of changes required, but as this task continues to expand, it is becoming more and more reasonable that we may as well clean up this sort of stuff (which seems to exists in each and everyone of the modes, BTW).